nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/state/server/ZooKeeperStateServer.java

```diff
@@ -198,6 +219,144 @@ public static ZooKeeperStateServer create(final NiFiProperties properties) throw
+                                  "nifi.zookeeper.client.secure was set to false. Check the administration guide.",
+                                   niFiProperties.getProperty(NiFiProperties.STATE_MANAGEMENT_ZOOKEEPER_PROPERTIES)));
+        }
+
+        // Remove any insecure ports if they were set in zookeeper.properties
+        ensureOnlySecurePortsAreEnabled(peerConfig, zkProperties);
+
+        // Set base ZooKeeper TLS server properties
+        setTlsProperties(zkProperties, new ZooKeeperServerX509Util(), niFiProperties);
+        // Set quorum ZooKeeper TLS server properties
+        setTlsProperties(zkProperties, new ZooKeeperQuorumX509Util(), niFiProperties);
+        // Set TLS client port:
+        zkProperties.setProperty("secureClientPort", getSecurePort(peerConfig));
+
+        // Set the required connection factory for TLS
+        zkProperties.setProperty(ZOOKEEPER_SERVER_CNXN_FACTORY, NettyServerCnxnFactory.class.getName());
+        zkProperties.setProperty(ZOOKEEPER_SSL_QUORUM, Boolean.TRUE.toString());
+
+        // Port unification allows both secure and insecure connections - setting to false means only secure connections will be allowed.
+        zkProperties.setProperty(ZOOKEEPER_PORT_UNIFICATION, Boolean.FALSE.toString());
+
+        // Recreate and reload the adjusted properties to ensure they're still valid for ZK:
+        peerConfig = new QuorumPeerConfig();
+        peerConfig.parseProperties(zkProperties);
+        return peerConfig;
+    }
+
+    private static boolean isZooKeeperConfigSecure(QuorumPeerConfig peerConfig) throws ConfigException {
+        InetSocketAddress secureAddress = peerConfig.getSecureClientPortAddress();
+        InetSocketAddress insecureAddress = peerConfig.getClientPortAddress();
+
+        if(secureAddress == null && insecureAddress == null) {
+            throw new ConfigException("No clientAddress or secureClientAddress is set in zookeeper.properties");
+        }
+
+        return secureAddress != null;
+    }
+
+    /**
+     * Verify whether the NiFi properties portion of ZooKeeper are correctly configured for TLS or not
+     * @param niFiProperties The loaded nifi.properties
+     * @return True if NiFi has TLS configuration and the property nifi.zookeeper.client.secure=true, otherwise false or configuration exception
+     * @throws ConfigException If nifi.zookeeper.client.secure=true but no TLS configuration is present
+     */
+    private static boolean isNiFiConfigSecureForZooKeeper(NiFiProperties niFiProperties) throws ConfigException {
+        final boolean istlsConfigPresent = niFiProperties.isZooKeeperTlsConfigurationPresent() || niFiProperties.isTlsConfigurationPresent();
+        final boolean isZooKeeperClientSecure = Boolean.parseBoolean(niFiProperties.getProperty(NiFiProperties.ZOOKEEPER_CLIENT_SECURE));
```

Comment:
There's a `isZooKeeperClientSecure()` in `NiFiProperties` to avoid unexpected behavior with `parseBoolean()`
```suggestion
        final boolean isTlsConfigPresent = niFiProperties.isZooKeeperTlsConfigurationPresent() || niFiProperties.isTlsConfigurationPresent();
        boolean isZooKeeperClientSecure = niFiProperties.isZooKeeperClientSecure();
```